src/uct/tcp/tcp_ep.c

```diff
@@ -290,15 +309,41 @@ void uct_tcp_ep_destroy(uct_ep_h tl_ep)
     }
 }
 
-void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
+static unsigned uct_tcp_ep_failed_progress(void *arg)
 {
+    uct_tcp_ep_t *ep       = (uct_tcp_ep_t*)arg;
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
 
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_FAILED);
+
     uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CLOSED);
     uct_set_ep_failed(&UCS_CLASS_NAME(uct_tcp_ep_t),
                       &ep->super.super, &iface->super.super,
                       UCS_ERR_ENDPOINT_TIMEOUT);
+
+    return 1;
+}
+
+void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface   = ucs_derived_of(ep->super.super.iface,
+                                              uct_tcp_iface_t);
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+
+    if (ep->flags & UCT_TCP_EP_FLAG_FAILED) {
+        /* error handling callback has already been scheduled and socket
+         * fd has been closed, no need to do these actions one more time */
+        ucs_assert(ep->fd == -1);
+        return;
+    }
+
+    uct_tcp_ep_mod_events(ep, 0, ep->events);
+    ucs_close_fd(&ep->fd);
```

Comment:
do we really have to close fd here?

Reply:
we could not do this
then it will be closed in the EP object destructor

Reply:
done